### ISSUE TITLE
False "Loop will execute at most once" warning emited for iterator function

### ISSUE THREAD

**Issue by Aspallar** [mpeterv/luacheck#153]
```
local Land = {}
local Creature = {}
local Artifact = {}
local Enchantment = {}
local Instant = {}
local Sorcery = {}
local Planeswalker = {}

-- luacheck: push ignore 512
-- lucheck falsely emits a "Loop will execute at most once" warning
local function AllTypelistEntries()
    local typelists = {Land, Creature, Artifact, Enchantment, Instant, Sorcery, Planeswalker}
    local entryIndex = 1
    local typelistIndex = 1
    return function ()
        while (typelistIndex <= #typelists) do
            local typelist = typelists[typelistIndex]
            while (entryIndex <= #typelist) do
                local entry = typelist[entryIndex]
                entryIndex = entryIndex + 1
                return entry
            end
            typelistIndex = typelistIndex + 1
            entryIndex = 1
        end
    end
end
-- luacheck: pop
```
Happens on the "while (entryIndex <= #typelist) do" line.

> luacheck --version
Luacheck: 0.21.2
Lua: Lua 5.3
LuaFileSystem: LuaFileSystem 1.6.3
LuaLanes: 3.10.1

**Comment by mpeterv** [comment]
Hello!

That's not a false positive, the loop in question:

```lua
while (entryIndex <= #typelist) do
    local entry = typelist[entryIndex]
    entryIndex = entryIndex + 1
    return entry
end
```

has an unconditional `return`, so when it is entered, the loop body is never repeated more than once, and the loop can be replaced with an `if` statement:

```lua
if (entryIndex <= #typelist) then
    local entry = typelist[entryIndex]
    entryIndex = entryIndex + 1
    return entry
end
```

**Comment by Aspallar** [comment]
Ah I see. I stand corrected. Thanks for the informative reply,
